sherpa/astro/ui/tests/test_astro_session.py

```diff
@@ -406,3 +413,815 @@ def test_save_error_ascii(session, kwargs, idval, tmp_path):
     assert len(data) == 2
     assert data[0] == pytest.approx([1, 3, 5])
     assert data[1] == pytest.approx([1, 2, 6])
+
+
+@pytest.mark.parametrize("session,flag",
+                         [(Session, False),
+                          (AstroSession, True)])
+def test_show_data(session, flag):
+    """Is show_data doing anything sensible?"""
+
+    s = session()
+
+    s.load_arrays(1, [-300, -100], [-150, -50], [-200, -100], Data1DInt)
+    s.load_arrays(2, [1, 2, 10, 20], [3, 5, 12, 20])
+
+    out = StringIO()
+    s.show_data(outfile=out)
+
+    toks = out.getvalue().split("\n")
+    print(out.getvalue())
+    assert toks[0] == "Data Set: 1"
+    idx = 1
+    if flag:
+        assert toks[idx] == "Filter: -300.0000--50.0000 x"
+        idx += 1
+
+    assert toks[idx] == "name      = "
+    idx += 1
+    assert toks[idx] == "xlo       = Int64[2]"
+    idx += 1
+    assert toks[idx] == "xhi       = Int64[2]"
+    idx += 1
+    assert toks[idx] == "y         = Int64[2]"
+    idx += 1
+    assert toks[idx] == "staterror = None"
+    idx += 1
+    assert toks[idx] == "syserror  = None"
+    idx += 1
+    assert toks[idx] == ""
+    idx += 1
+    assert toks[idx] == "Data Set: 2"
+    idx += 1
+
+    if flag:
+        assert toks[idx] == "Filter: 1.0000-20.0000 x"
+        idx += 1
+
+    assert toks[idx] == "name      = "
+    idx += 1
+    assert toks[idx] == "x         = Int64[4]"
+    idx += 1
+    assert toks[idx] == "y         = Int64[4]"
+    idx += 1
+    assert toks[idx] == "staterror = None"
+    idx += 1
+    assert toks[idx] == "syserror  = None"
+    idx += 1
+    assert toks[idx] == ""
+    idx += 1
+    assert toks[idx] == ""
+    idx += 1
+    assert toks[idx] == ""
+
+    n = 19 if flag else 17
+    assert len(toks) == n
+
+
+@pytest.mark.parametrize("session", [Session, AstroSession])
+def test_show_method(session):
+    """Is show_method doing anything sensible?"""
+
+    s = session()
+
+    # Select the method so we are not affected if the default changes.
+    #
+    s.set_method("levmar")
+
+    out = StringIO()
+    s.show_method(outfile=out)
+
+    toks = out.getvalue().split("\n")
+    assert toks[0] == "Optimization Method: LevMar"
+    assert toks[1] == "name     = levmar"
+    assert toks[2] == "ftol     = 1.1920928955078125e-07"
```

Comment:
I'm glad this works, but I'm always wary of hardcoding so much precision. It's prone to failing and creating confusion (e.g. if some user runs it on some other hardware or python version) or on totally unrelated changes because some default numpy printing changed. I suggest to check only what is the minimum we expect for this to make sense, e.g. `assert re.match("1.19[0-9]*e-07")`. We want to test to fail is this is much more or less than 1e-7 or of the formatting of the print statement changes significantly, but we don't need it to fail because the 17th digit is either not printed to differs in some way.

While we have some tests that are classic regression tests ("Make sure we manually check every change"), here I would advocate for a setup that's a little more lenient. If this was not a string comparison, we would use `pytest.approx` for the same effect.

Reply:
We have quite a few tests that I expect to fail at some point if there's significant numerical differences. I've actually been quite surprised we haven't seen many already. I think that it's in part many of these values (like the 1.19..e-7) value are "special" - e.g.

```
>>> numpy.finfo(numpy.float32).eps
1.1920929e-07
```

and so they don't suffer from "'rounding/..." issues that a value that was calculated would have. There's still the issue of how-many decimal places get displayed, but I think that is either a python or numpy decision (depending on exactly how it's created).

So in this case I don't really have the energy to change these lines until we start seeing a problem with them, but there are other places where we could be more defensive.